Normalize sampled MCDA weights

Rosejoycrocker opened the pull request:
Incorporates changes allowing MCDA weights for fogging and seeding decision matrices to be normalised at the parameter sampling step so that parameter inputs reflect their magnitude as used inside ADRIA.

The heat stress and wave stress weights are also split into separate weights for seeding and fogging so that normalisation can be done correctly (otherwise normalised values for seeding will be overwritten by normalised values for shading and these will not be the same).

Tests added to check normalisation is working as expected.

Note the growth.jl test is failing but this is fixed in PR #573

Closes #559.

Closes #590

ConnectedSystems commented on src/decision/CriteriaWeights.jl:
```diff
@@ -1,28 +1,51 @@
+        name="Seed Heat Stress",
+        description="Importance of avoiding heat stress when seeding. Higher values places more weight on areas with low heat stress.",
+    )
+    shade_wave_stress::P = Param(
+        1.0;
+        ptype="real",
+        bounds=(0.0, 1.0),
+        default_bounds=(0.0, 1.0),
+        dists="unif",
+        criteria_keywords=(:fog, :weight),
+        name="Shade Wave Stress",
+        description="Importance of avoiding wave stress when fogging. Higher values places more weight on areas with low wave stress.",
+    )
+    shade_heat_stress::P = Param(
+        1.0;
+        ptype="real",
+        bounds=(0.0, 1.0),
+        default_bounds=(0.0, 1.0),
+        dists="unif",
+        criteria_keywords=(:fog, :weight),
+        name="Shade Heat Stress",
+        description="Importance of avoiding heat stress when fogging. Higher values places more weight on areas with low heat stress.",
     )
```
Description and criteria keywords mention `fog`, but factor and field names use "shade" ?

Rosejoycrocker replied:
Yeah, I originally had "shade" as the descriptor but really it's fogging so I changed it. I can change it back to shade for consistency, or I can change the parameters labeled with "shade" to be labeled with "fog", which will require some other changes throughout the code. What do you think?

ConnectedSystems replied:
Which is more specific to what is being represented?

Rosejoycrocker replied:
"fog" is more relevant, I can change it to that. Is it ok to also change the order of some parameter names in this PR while I'm at it? Some have seed/shade at the end, some at the start etc.

ConnectedSystems replied:
Yes?

ConnectedSystems commented on src/decision/CriteriaWeights.jl:
```diff
@@ -1,28 +1,51 @@
         dists="unif",
-        name="Wave Stress",
-        description="Importance of avoiding wave stress. Higher values places more weight on areas with low wave stress.",
+        criteria_keywords=(:seed, :weight),
```
Could you elaborate on what the `:weight` keyword is used for? The struct is ostensibly for criteria weights, so are we adding a `:weight` keyword for all parameters here?

ConnectedSystems replied:
I gather it's to distinguish with `threshold` values - okay leave this as is, we'll clean up later as part of a larger restructure.

ConnectedSystems commented on src/ecosystem/corals/Corals.jl:
```diff
@@ -296,7 +296,10 @@ function create_coral_struct(bounds::Tuple{Float64,Float64}=(0.9, 1.1))::Nothing
         default_bounds=(0.25, 0.5),
         dists="unif",
-        name="Heritability", description="Heritability of DHW tolerance.")
+        criteria_keywords=(""),
+        name="Heritability",
+        description="Heritability of DHW tolerance.",
+    )
```
A bit late now but you could have created your own separate constructor that defines a default value for these additional fields (`criteria_keywords`, `default_bounds`, etc) so you don't have to go around adding these fields to all use of `Param` types.

Leave it as is - we'll clean it up later.

ConnectedSystems commented on src/io/sampling.jl:
```diff
@@ -43,6 +46,15 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
     not_fogged = (df.fogging .== 0) .& (df.SRM .== 0)
     df[not_fogged, contains.(names(df), "shade_")] .= 0.0
 
+    # Normalize MCDA weights for fogging scenarios
+    df[.!(not_fogged) .& (df.guided .> 0), weights_fog_crit.fieldname] .= mcda_normalize(
+        df[.!(not_fogged) .& (df.guided .> 0), weights_fog_crit.fieldname],
+    )
```
SRM is applied on a regional basis at the moment - shouldn't `not_fogged` just check if `fogging` == 0?

Rosejoycrocker replied:
good point!

ConnectedSystems reviewed (changes requested):
Just one question from me.

Rosejoycrocker commented:
I think I've addressed your comments @ConnectedSystems . Tests are all passing on my end. For consistency with changing "shade" to "fog" in `CriteriaWeights.jl` I also changed "shade" to "fog" in other relevant places inside the MCDA, such as the weight names and `pref_shade_sites`. `shade_locs` is also now `fog_locs` in `scenario.jl` and I renamed `create_shade_matrix` to `create_fog_matrix`.

ConnectedSystems commented on src/io/sampling.jl:
```diff
@@ -41,7 +44,16 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
     # Same for fogging/shading
     not_fogged = (df.fogging .== 0) .& (df.SRM .== 0)
-    df[not_fogged, contains.(names(df), "shade_")] .= 0.0
+    df[not_fogged, contains.(names(df), "fog_") .& contains.(names(df), "shade_")] .= 0.0
+
+    # Normalize MCDA weights for fogging scenarios
+    df[(df.fogging .> 0.0) .& (df.guided .> 0), weights_fog_crit.fieldname] .= mcda_normalize(
+        df[(df.fogging .> 0.0) .& (df.guided .> 0), weights_fog_crit.fieldname],
+    )
+    # Normalize MCDA weights for seeding scenarios
+    df[.!(not_seeded) .& (df.guided .> 0), weights_seed_crit.fieldname] .= mcda_normalize(
+        df[.!(not_seeded) .& (df.guided .> 0), weights_seed_crit.fieldname],
+    )
```
Is this the most readable approach?

Rosejoycrocker commented:
I thinkI've addressed your comments @ConnectedSystems

ConnectedSystems reviewed (changes requested):
Sorry, I forgot to hit the submit button for these comments

Rosejoycrocker commented on src/decision/dMCDA.jl:
```diff
@@ -182,6 +187,14 @@ function mcda_normalize(x::Matrix)::Matrix
+"""
+function mcda_normalize(x::DataFrame)::DataFrame
+    return x ./ sum.(eachrow(x))
+end
```
Weird, tests are passing for me. WHat you've linked looks different from the current implementation- have you pulled the most recent version?

ConnectedSystems commented on src/interventions/Interventions.jl:
```diff
@@ -117,6 +129,7 @@ Base.@kwdef struct Intervention{N,P,P2} <: EcoModel
         bounds=(2.0, 25.0 + 1.0),
         default_bounds=(2.0, 25.0 + 1.0),
         dists="unif",
+        criteria_keywords=(""),
```
You're aware that this creates an empty string?

If you want a tuple with a single element, it is `("", )`

ConnectedSystems commented on src/io/sampling.jl:
```diff
@@ -41,7 +44,17 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
 
     # Same for fogging/shading
     not_fogged = (df.fogging .== 0) .& (df.SRM .== 0)
-    df[not_fogged, contains.(names(df), "shade_")] .= 0.0
+    df[not_fogged, contains.(names(df), "fog_") .& contains.(names(df), "shade_")] .= 0.0
+
+    # Normalize MCDA weights for fogging scenarios
+    is_guided = df.guided .> 0
+    df[(df.fogging .> 0.0) .& is_guided, weights_fog_crit.fieldname] .= mcda_normalize(
+        df[(df.fogging .> 0.0) .& is_guided, weights_fog_crit.fieldname],
+    )
+    # Normalize MCDA weights for seeding scenarios
+    df[.!(not_seeded) .& is_guided, weights_seed_crit.fieldname] .= mcda_normalize(
+        df[.!(not_seeded) .& is_guided, weights_seed_crit.fieldname],
+    )
```
Line 46 still needs to be addressed.

I also don't consider the DataFrame selectors to be very readable but if you're okay with it then I'll live.

Rosejoycrocker replied:
I've changed them again

ConnectedSystems commented on src/decision/dMCDA.jl:
```diff
@@ -182,6 +187,15 @@ function mcda_normalize(x::Matrix)::Matrix
+
+Normalize weights for a set of scenarios (wse/wsh) for MCDA.
+"""
+function mcda_normalize(x::DataFrame)::DataFrame
+    norm_x = sum.(eachrow(x))
+    return x ./ norm_x
+end
```
Did you compare this with the previous approach converting to matrix?

Rosejoycrocker replied:
The Matrix version took 0.406 seconds, the eachrow() version took 0.082 seconds.

As I said in my email, I think this difference is due to the converison to Matrix taking time. I'm guessing eachrow is also optimised for use with dataframes.

ConnectedSystems replied:
How did you do the timing and did you check allocations?

Rosejoycrocker replied:
I assigned both versions to a function and did @time func(x)... I know its preferable to use Benchmarking Tools but I just did it quickly

ConnectedSystems replied:
If you're okay with that then let's roll with it

Rosejoycrocker replied:
I'm okay with it, but are you?

ConnectedSystems commented on src/io/sampling.jl:
```diff
@@ -41,7 +44,18 @@ function adjust_samples(spec::DataFrame, df::DataFrame)::DataFrame
+        df[guided_fogged, weights_fog_crit.fieldname],
+    )
+    # Normalize MCDA weights for seeding scenarios
+    guided_seeded = .!(not_seeded) .& is_guided
```
`is_guided` does not exist any more.

Tests fail due to this.

Rosejoycrocker replied:
Fixed now, sorry, and tests are passing on my end

ConnectedSystems reviewed (approved):
Tests pass after rebase and scenario runs and figure creation work without issue.
